File: python-advent-of-code/advent/twenty_fifteen/day02_I_was_told_there_would_be_no_math.py
import re

from advent.utilities.file_reader import FileReader


class Day02IWasToldThereWouldBeNoMath:
    def __init__(self, filename):
        presents = FileReader.read_file(filename, Day02IWasToldThereWouldBeNoMath.present_converter)
        self.part_1_answer = sum([Day02IWasToldThereWouldBeNoMath.calc_1(present) for present in presents])
        self.part_2_answer = sum([Day02IWasToldThereWouldBeNoMath.calc_2(present) for present in presents])

    @staticmethod
    def calc_1(present):
        a = present[0]
        b = present[1]
        c = present[2]
        return 3 * a * b + 2 * b * c + 2 * a * c

    @staticmethod
    def calc_2(present):
        a = present[0]
        b = present[1]
        c = present[2]
        return 2 * a + 2 * b + a * b * c

    @staticmethod
    def present_converter(line):
        m = re.match(r'([0-9]*)x([0-9]*)x([0-9]*)', line)
        r = [int(m.group(1)), int(m.group(2)), int(m.group(3))]
        r.sort()
        return r

File: scala-advent-of-code/src/main/scala/advent/twenty_fifteen/Day02IWasToldThereWouldBeNoMath.scala
package advent.twenty_fifteen

import advent.common.DailyProblem
import advent.utilities.FileUtilities

case class Parcel(length: Int, width: Int, height: Int)

class Day02IWasToldThereWouldBeNoMath(filename: String) extends DailyProblem[Int, Int] {
  private val parcels = FileUtilities.readFile(filename, parse)

  private def parse(line: String): Parcel = {
    val bits = line.split("x")
    Parcel(bits(0).toInt, bits(1).toInt, bits(2).toInt)
  }

  private def surfaceArea(p: Parcel): Int = 2 * p.length * p.width + 2 * p.width * p.height + 2 * p.height * p.length

  private def surfaceAreaOfSmallestSide(p: Parcel): Int = Math.min(p.height * p.width, Math.min(p.height * p.length, p.length * p.width))

  private def lengthToWrap(p: Parcel) = Math.min(2 * p.height + 2 * p.width, Math.min(2 * p.height + 2 * p.length, 2 * p.length + 2 * p.width))

  private def lengthOfBow(p: Parcel) = p.width * p.length * p.height

  override val part1Answer: Int = parcels.foldLeft(0)((a, p) => a + surfaceArea(p) + surfaceAreaOfSmallestSide(p))
  override val part2Answer: Int = parcels.foldLeft(0)((a, p) => a + lengthToWrap(p) + lengthOfBow(p))
}
